src/node_modules.cc

```diff
@@ -579,6 +579,88 @@ void GetCompileCacheEntry(const FunctionCallbackInfo<Value>& args) {
       isolate, v8::Null(isolate), names.data(), values.data(), names.size()));
 }
 
+static void InitImportMetaLazyGetter(
+    Local<v8::Name> name, const v8::PropertyCallbackInfo<Value>& info) {
+  Isolate* isolate = info.GetIsolate();
+  // This getter has no JavaScript function representation and is not
+  // invoked in the creation context.
+  // When this getter is invoked in a vm context, the `Realm::GetCurrent(info)`
+  // returns a nullptr and. Retrieve the creation context via `this` object and
+  // get the creation Realm.
+  Local<Value> receiver_val = info.This();
+  if (!receiver_val->IsObject()) {
+    THROW_ERR_INVALID_INVOCATION(isolate);
+    return;
+  }
+  Local<Object> receiver = receiver_val.As<Object>();
+  Local<Context> context;
+  if (!receiver->GetCreationContext().ToLocal(&context)) {
+    THROW_ERR_INVALID_INVOCATION(isolate);
+    return;
+  }
+  Environment* env = Environment::GetCurrent(context);
+
+  node::Utf8Value url(isolate, info.Data());
+  auto file_url = ada::parse(url.ToStringView());
+  CHECK(file_url);
+  auto file_path = url::FileURLToPath(env, *file_url);
+  CHECK(file_path.has_value());
+  std::string_view ret_view = file_path.value();
+
+  node::Utf8Value utf8name(isolate, name);
+  auto plain_name = utf8name.ToStringView();
+  if (plain_name == "dirname") {
+#ifdef _WIN32
+#define PATH_SEPARATOR '\\'
+#else
+#define PATH_SEPARATOR '/'
+#endif
+    auto index = ret_view.rfind(PATH_SEPARATOR);
+#undef PATH_SEPARATOR
+    CHECK(index != std::string_view::npos);
+    ret_view.remove_suffix(ret_view.size() - index);
+  }
+  Local<Value> ret;
+  if (ToV8Value(context, ret_view, isolate).ToLocal(&ret)) {
+    info.GetReturnValue().Set(ret);
+  }
+}
+void InitImportMeta(const FunctionCallbackInfo<Value>& args) {
+  // target, url, shouldSetDirnameAndFilename, resolve
+  CHECK_GE(args.Length(), 4);
+  CHECK(args[0]->IsObject());
+  CHECK(args[1]->IsString());
+  CHECK(args[2]->IsBoolean());
+  CHECK(args[3]->IsFalse() || args[3]->IsFunction());
+
+  Isolate* isolate = args.GetIsolate();
+  Local<Context> context = isolate->GetCurrentContext();
+
+  auto target = args[0].As<Object>();
+
+  // N.B.: Order is important to keep keys in alphabetical order.
+
+  if (args[2]->IsTrue()) {
+    target
+        ->SetLazyDataProperty(context,
+                              FIXED_ONE_BYTE_STRING(isolate, "dirname"),
+                              InitImportMetaLazyGetter,
+                              args[1])
+        .Check();
+    target
+        ->SetLazyDataProperty(context,
+                              FIXED_ONE_BYTE_STRING(isolate, "filename"),
+                              InitImportMetaLazyGetter,
+                              args[1])
+        .Check();
+  }
+
+  if (args[3]->IsFunction()) {
+    target->Set(context, FIXED_ONE_BYTE_STRING(isolate, "resolve"), args[3])
+        .Check();
+  }
+  target->Set(context, FIXED_ONE_BYTE_STRING(isolate, "url"), args[1]).Check();
+}
```

Comment:
We could use env properties.

```suggestion
  if (args[3]->IsFunction()) {
    target->Set(context, env->resolve_string(), args[3])
        .Check();
  }
  target->Set(context, env->url_string(), args[1]).Check();
}
```

Would need to retrieve `env`. Maybe with `Environment* env = Environment::GetCurrent(args);`.

And would need to add `"resolve"` to https://github.com/nodejs/node/blob/main/src/env_properties.h